### DISCUSSION TITLE
Create an official ESLint plugin

### DISCUSSION THREAD

**Discussion by aaronadamsCA** [remix-run/react-router#12522] [Proposals]
I think it would be a great idea to ship an official `@react-router/eslint-plugin`.

Similar to other projects' official plugins like [`@vitest/eslint-plugin`](https://github.com/vitest-dev/eslint-plugin-vitest) and [`eslint-plugin-storybook`](https://github.com/storybookjs/eslint-plugin-storybook), this plugin would contain rules that help developers of all skill levels avoid mistakes.

A few rule ideas to start:

- `no-action-only-routes`: Require a route that exports an action to also export a loader or a route component.

  If an action-only route returns an error and the user refreshes the page, React Router returns an error because the route cannot respond to `GET` requests. The simplest fix is to add a loader that redirects the user to a valid path.

- `require-root-error-boundary`: Make sure the root route exports an error boundary.

- `valid-resource-route`: Make sure a resource root returns a `Response`.

- `valid-route-module-path`: Validate that the route module paths in `routes.ts` can be resolved (or better yet, that they resolve to valid route modules).

Like any good ESLint plugin, the simplest rules would help to reduce the learning curve (and the corresponding support load), while the most advanced rules could help large apps find and fix all kinds of edge-case problems.

Presumably there could also be shared configs, e.g. a "recommended" config that checks for common errors, and a "strict" config that extends the recommended config with a more opinionated set of best practices.

**Comment by Tristan-WorkGH** [comment]
Another idea: a rule to forbid the usage of components, functions or attributes that can't be used in the declarative or data mode.  
https://reactrouter.com/start/modes#api--mode-availability-table

**Comment by brophdawg11** [comment]
We chatted about this in our Steering Committee meeting today and we don't think this is something we're interested in as a core package.  Linting preferences vary greatly team to team and if we shipped something for eslint, it might start small, but eventually we'd have continued requests for new rules and variations of those rules to fit specific team preferences.  It also likely wouldn't be long before folks started to request equivalents for biome, oxlint, etc.  It feels like it could easily become something that takes our focus away from working on react-router itself. 

I think this would make a great community package though 😄

**Comment by apexskier** [comment]
Another idea, if someone makes one, is ensuring the correct `Route` type is imported:

Bad:

```ts
import type { Route } from "../+types/foo";
```
